File: packages/draftfirst/src/normalize.ts
/**
 * Idempotent commit-time normalization for parentheticals and cue extensions.
 * Other text is returned unchanged.
 */

import type { ElementType } from './types.js';

/** Canonical cue extensions, keyed by their letters with punctuation stripped. */
const EXTENSION_CANONICAL: Record<string, string> = {
	VO: 'V.O.',
	OS: 'O.S.',
	OC: 'O.C.',
	CONTD: "CONT'D",
	PRELAP: 'PRE-LAP',
	SUBTITLE: 'SUBTITLE',
	FILTERED: 'FILTERED'
};

// Bare trailing extension: whitespace + the extension letters (dots optional)
// at end of string. The whitespace boundary keeps names like CARLOS safe.
const BARE_EXTENSION = /\s+(V\.?O\.?|O\.?S\.?|O\.?C\.?|CONT'?D\.?|PRE[\s-]?LAP|SUBTITLE|FILTERED)$/i;

function canonicalExtension(raw: string): string | null {
	const key = raw.toUpperCase().replace(/[.\s'-]/g, '');
	return EXTENSION_CANONICAL[key] ?? null;
}

/**
 * Wrap a parenthetical's text in exactly one pair of brackets.
 * Strips one layer of partial outer brackets first, so partial states the
 * writer typed ('(beat', 'beat)', '()') all converge. Idempotent.
 */
export function normalizeParenthetical(text: string): string {
	let s = text.trim();
	if (s.startsWith('(')) s = s.slice(1);
	if (s.endsWith(')')) s = s.slice(0, -1);
	s = s.trim();
	return s ? `(${s})` : '';
}

/**
 * Normalize a character cue's extension:
 *   'MARA (V.O'        → 'MARA (V.O.)'   (close + canonicalize)
 *   'MARA (WHISPERING' → 'MARA (WHISPERING)' (close any unclosed extension)
 *   'MARA ('           → 'MARA'          (drop a dangling empty bracket)
 *   'MARA VO'          → 'MARA (V.O.)'   (bracket a bare known extension)
 *   'MARA (V.O.)'      → unchanged
 *   'VO'               → unchanged       (an extension with no name is a name)
 */
export function normalizeCue(text: string): string {
	const s = text.trim();
	if (!s) return '';

	// Preserve already-closed extensions.
	if (s.endsWith(')')) return s;

	const open = s.indexOf('(');
	if (open >= 0) {
		const name = s.slice(0, open).trim();
		const tail = s.slice(open + 1).trim();
		if (!tail) return name; // dangling '(' — drop it
		if (!name) return s; // '(' at the very start — not a cue shape, leave it
		const canon = canonicalExtension(tail);
		return `${name} (${canon ?? tail})`;
	}

	// No bracket at all — a bare known extension gets brackets.
	const m = s.match(BARE_EXTENSION);
	if (m) {
		const name = s.slice(0, m.index).trim();
		const canon = canonicalExtension(m[1]);
		if (name && canon) return `${name} (${canon})`;
	}

	return s;
}

/**
 * A wryly lane holding an ALL-CAPS name is not a wryly — it is a cue that
 * took the wrong lane (after dialogue, Tab lands on parenthetical, and a
 * writer thinking about the next speaker may not notice the indent).
 * Wrylies are lowercase by convention; cues are uppercase. The shape test
 * is deliberately strict — mixed-case direction like '(to JOHN)' is never
 * hijacked, and neither is a genuine shouted wryly typed with a bracket.
 */
export function looksLikeCue(text: string): boolean {
	const t = text.trim();
	if (t.length < 2 || t.startsWith('(')) return false;
	if (!/[A-Za-z]/.test(t)) return false;
	return t === t.toUpperCase();
}

/**
 * Commit-time normalization for a block's text, dispatched by element type.
 * Scene / action / dialogue / transition are returned untouched.
 */
export function normalizeElementText(type: ElementType, text: string): string {
	if (type === 'parenthetical') return normalizeParenthetical(text);
	if (type === 'character') return normalizeCue(text);
	return text;
}

File: ios/eDraftEngine/Sources/EDraftEngine/Editor/Normalize.swift
import Foundation

/// Text normalization rules for character cues and parentheticals, ported
/// from the TypeScript engine's `normalize.ts`. Behaviour is pinned by
/// `Fixtures/normalize.json` — including the deliberate ASCII-only regex
/// semantics (a curly apostrophe `'` in `CONT'D` does NOT key the canonical
/// extension table, exactly as in JS).
public enum Normalize {

    // MARK: - Character extensions

    /// Canonical spellings keyed by the stripped, uppercased tail
    /// (TypeScript `EXTENSION_CANONICAL`).
    private static let extensionCanonical: [String: String] = [
        "VO": "V.O.",
        "OS": "O.S.",
        "OC": "O.C.",
        "CONTD": "CONT'D",
        "PRELAP": "PRE-LAP",
        "SUBTITLE": "SUBTITLE",
        "FILTERED": "FILTERED",
    ]

    /// Strips `[.\s'-]` — ASCII apostrophe only, matching the JS character
    /// class exactly (a curly `'` is preserved). Built locally because
    /// `Regex` is not `Sendable` and cannot be a static under Swift 6
    /// concurrency checking.
    private static func canonicalExtension(_ raw: String) -> String? {
        let strip = #/[.\s'-]/#
        let key = raw.replacing(strip, with: "").uppercased()
        return extensionCanonical[key]
    }

    // MARK: - Public API

    /// TypeScript `normalizeParenthetical`: trim, strip ONE leading `(` and
    /// ONE trailing `)`, trim, then wrap in `()` when non-empty.
    public static func normalizeParenthetical(_ text: String) -> String {
        var s = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.hasPrefix("(") { s.removeFirst() }
        if s.hasSuffix(")") { s.removeLast() }
        s = s.trimmingCharacters(in: .whitespacesAndNewlines)
        return s.isEmpty ? "" : "(\(s))"
    }

    /// TypeScript `unwrapParenthetical`: the inverse, for conversion OUT of
    /// the parenthetical lane. Sheds exactly one outer wrapper — "(beat)" →
    /// "beat", "((beat))" → "(beat)". A text holding several directions,
    /// "(a) (b)", keeps them: only a first "(" that closes at the very end
    /// is a wrapper. Partial states shed one stray end bracket; anything
    /// unbalanced is left exactly as written.
    public static func unwrapParenthetical(_ text: String) -> String {
        let s = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.isEmpty { return "" }
        let opens = s.contains("(")
        let closes = s.contains(")")
        if !opens && !closes { return s }
        if !opens {
            // One stray closer: "beat)" → "beat" (one layer only).
            guard let i = s.lastIndex(of: ")") else { return s }
            var t = s
            t.remove(at: i)
            return t.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if !closes {
            // One stray opener: "(beat" → "beat" (one layer only).
            guard let i = s.firstIndex(of: "(") else { return s }
            var t = s
            t.remove(at: i)
            return t.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        // Both present: strip only a true outer wrapper.
        var depth = 0
        for (offset, ch) in s.enumerated() {
            if ch == "(" { depth += 1 }
            if ch == ")" {
                depth -= 1
                if depth < 0 { return s }
                if depth == 0 {
                    guard offset == s.count - 1 else { return s }
                    return String(s.dropFirst().dropLast())
                        .trimmingCharacters(in: .whitespacesAndNewlines)
                }
            }
        }
        return s
    }

    /// TypeScript `normalizeCue`: canonicalize the extension on a character
    /// cue — `MARA (V.O)`, `MARA VO`, `MARA (vo)` all become `MARA (V.O.)`.
    public static func normalizeCue(_ text: String) -> String {
        // Bare trailing extension detector (TypeScript `BARE_EXTENSION`,
        // JS `/i`). Local because `Regex` is not `Sendable`.
        let bareExtension = #/(?i)\s+(V\.?O\.?|O\.?S\.?|O\.?C\.?|CONT'?D\.?|PRE[\s-]?LAP|SUBTITLE|FILTERED)$/#
        let s = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.hasSuffix(")") { return s }                    // already bracketed
        guard let open = s.firstIndex(of: "(") else {
            // No bracket: detect a bare trailing extension.
            if let match = s.firstMatch(of: bareExtension) {
                let name = String(s[s.startIndex..<match.range.lowerBound])
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                guard !name.isEmpty else { return s }
                let tail = String(match.output.1)
                let canonical = canonicalExtension(tail) ?? tail.uppercased()
                return "\(name) (\(canonical))"
            }
            return s
        }
        let name = String(s[s.startIndex..<open])
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let tail = String(s[s.index(after: open)...])
            .trimmingCharacters(in: .whitespacesAndNewlines)
        if tail.isEmpty { return name }
        if name.isEmpty { return s }
        let canonical = canonicalExtension(tail) ?? tail
        return "\(name) (\(canonical))"
    }

    /// TypeScript `looksLikeCue`: trimmed, ≥2 chars, not starting with `(`,
    /// contains an ASCII letter, and equal to its uppercased form.
    public static func looksLikeCue(_ text: String) -> Bool {
        let t = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard t.count >= 2, !t.hasPrefix("(") else { return false }
        // JS: /[A-Za-z]/ — ASCII letters only, deliberately.
        guard t.contains(where: { $0.isASCII && $0.isLetter }) else { return false }
        return t == t.uppercased()
    }

    /// TypeScript `normalizeElementText`: dispatch per element kind.
    public static func normalizeElementText(kind: ElementKind, text: String) -> String {
        switch kind {
        case .parenthetical: return normalizeParenthetical(text)
        case .character: return normalizeCue(text)
        default: return text
        }
    }
}
